### ISSUE TITLE
authUI:didSignInWithUser: not being called for FUIPhoneAuth - objective-c

### ISSUE THREAD

**Issue by wayneforrest** [firebase/FirebaseUI-iOS#310]
I am having some problems where my delegate method "didSignInWithUser" is not being called,
after I a have entered the "SMS Code" received.  

*not sure about this anymore * >>Does get called when user cancels.

## initiator:
`
SignUpController *su = [SignUpController new];
[su signUpWithController:self]; //self is a UIViewController
`

## implementation:
`
@import FirebasePhoneAuthUI;

@interface SignUpController ()

@property (weak,nonatomic) UIViewController *controller;
@property (strong,nonatomic) FUIAuth *auth;

@end

@implementation SignUpController

-(void)login {
    
    _auth = [FUIAuth defaultAuthUI];
    _auth.delegate = self;
    
    FUIPhoneAuth *phoneAuth = [[FUIPhoneAuth alloc] initWithAuthUI:_auth];
    _auth.providers = @[phoneAuth];
    [phoneAuth signInWithPresentingViewController:_controller];
}

-(void)signUpWithController:(UIViewController*)controller {
    
    _controller = controller;
    [self login];
}

-(void)authUI:(FUIAuth *)authUI didSignInWithUser:(FIRUser *)user error:(NSError *)error {
    
    if (error) {
        NSLog(@"error: %@",error.localizedDescription);
        [self login];
        return;
    }
    
    if (!user) {
        [self login];
    } else {
        
        // store user details
        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        [defaults setObject:user.uid forKey:@"user.uid"];
        [defaults setObject:user.providerID forKey:@"user.providerID"];
        [defaults setObject:user.phoneNumber forKey:@"user.phoneNumber"];
        
        // log user details
        NSLog(@"Signup success with:");
        NSLog(@" - uid:%@",user.uid);
        NSLog(@" - providerID:%@",user.providerID);
        NSLog(@" - phoneNumber:%@",user.phoneNumber);
    }
}
`

**Comment by wayneforrest** [comment]
I have tried changing the way the view controller is presented as recommended by some other posts,
but still my delegate method is not being called.

`-(void)login {
    
    _auth = [FUIAuth defaultAuthUI];
    _auth.delegate = self;
    
    FUIPhoneAuth *phoneAuth = [[FUIPhoneAuth alloc] initWithAuthUI:_auth];
    _auth.providers = @[phoneAuth];
    [phoneAuth signInWithPresentingViewController:_controller];
    
    UIViewController *vc = [_auth authViewController]; //returns your custom controller wrapped in UINavigationController
    [self.controller presentViewController:vc animated:YES completion:nil];
    
}`

**Comment by wayneforrest** [comment]
@jeandavid @yramanchuk Hi, I think you know how to solve this by looking at some of the previous related issues. When I look at the Web FireBase Console I can see that the user is registered, it is just that my delegate method does not get called, which I need to get the userId.

Maybe you can spot the mistake or send me some working example, Thanks.

**Comment by wayneforrest** [comment]
I have noticed the way the controller gets dismissed seems to be source of my problem.

Line 170 located here FirebaseUI-iOS/FirebaseAuthUI/FUIAuth.m
[presentingViewController dismissViewControllerAnimated:YES....

I presume this will actually remove my presentingViewcontroller and thus I will receive no
callbacks.

I guess I will have to move my delegate reference to be some other class. 
I am considering my delegate to be AppDelegate.

`
if (error) {
   [self invokeResultCallbackWithUser:user error:error];
   } else {
       if (presentingViewController.presentingViewController) {
           [presentingViewController dismissViewControllerAnimated:YES completion:^{
           [self invokeResultCallbackWithUser:user error:error];
       }];
   } else {
       [self invokeResultCallbackWithUser:user error:error];
   }

`

**Comment by wayneforrest** [comment]
I can confirm that by using the AppDelegate as the delegate for the FUIAuth I do get callbacks.

Some other problems I ran into...
my NavigationBar was black but the fix was to set the following before displaying it:
```
// statusbar color
    UINavigationBar.appearance.translucent = NO;
    UINavigationBar.appearance.barStyle = UIBarStyleBlack;
    UINavigationBar.appearance.barTintColor = UIColorFromRGB(ThemeDTO.sharedInstance.navigation_bar);
```

Also when starting the application I have an FIRAuthStateDidChangeListenerHandle
that calls 
```
- (void)updateUI:(FIRAuth * _Nonnull) auth withUser:(FIRUser *_Nullable) user {
}

```
But for some reason when I start my application a second time the phone number is null.
To remedy I hacked my registration process a bit and store the phonenumer in NSUserDefauts, when I first register the user.

**Comment by yramanchuk** [comment]
@wayneforrest 

> I can confirm that by using the AppDelegate as the delegate for the FUIAuth I do get callbacks.

I bet that you haven't retained delegate which is weak reference (so it was deallocated immediately). Similar issue: https://github.com/firebase/FirebaseUI-iOS/issues/294#issuecomment-307285100

> But for some reason when I start my application a second time the phone number is null.

Please refer to your issue https://github.com/firebase/FirebaseUI-iOS/issues/312#issuecomment-315149649

> my NavigationBar was black but the fix was to set the following before displaying it:

Please refer to https://github.com/firebase/FirebaseUI-iOS/issues/307#issuecomment-311519546

@wayneforrest Please let me know if you have any other issue.

**Comment by neilkachu** [comment]
Ive come across the issue with:

**Comment by yramanchuk** [comment]
I've fixed situation when the only available provider was shown from modal view controller.
In the described situation we were closing initial controller: it should be fixed in FirebaseUI 4.3.2
